compiler/src/vm/builtin_functions.rs

```diff
@@ -1,373 +1,443 @@
+                closure.address.format(self),
+            );
+            self.dup(closure.address);
+            DivergeControlFlow(closure.address)
         })
     }
 
-    fn if_else<U: UseProvider>(
-        &mut self,
-        use_provider: &U,
-        args: Vec<Value>,
-    ) -> Result<(), String> {
-        destructure!(
+    fn get_argument_count(&mut self, args: &[Pointer]) -> BuiltinResult {
+        unpack_and_later_drop!(self, args, (closure: Closure), {
+            Return(self.create_int(closure.num_args.into()))
+        })
+    }
+
+    fn if_else(&mut self, args: &[Pointer]) -> BuiltinResult {
+        unpack_and_later_drop!(
+            self,
             args,
-            [
-                Value::Symbol(condition),
-                Value::Closure(then_closure),
-                Value::Closure(else_closure)
-            ],
+            (condition: bool, then: Closure, else_: Closure),
             {
-                if then_closure.num_args > 0 {
-                    return Err(format!("IfElse expects a closure without arguments as the then, but got one with {} arguments.", then_closure.num_args));
-                }
-                if else_closure.num_args > 0 {
-                    return Err(format!("IfElse expects a closure without arguments as the else, but got one with {} arguments.", else_closure.num_args));
-                }
-                let condition = match condition.as_str() {
-                    "True" => true,
-                    "False" => false,
-                    _ => {
-                        return Err(format!(
-                            "IfElse expected True or False as a condition, but got {condition}.",
-                        ));
-                    }
-                };
-
-                let closure_object = self.heap.import(if condition {
-                    Value::Closure(then_closure.clone())
-                } else {
-                    Value::Closure(else_closure.clone())
-                });
-                log::debug!(
-                    "IfElse executing the closure: {:?}",
-                    self.heap.export_without_dropping(closure_object)
-                );
-                self.data_stack.push(closure_object);
-                self.run_instruction(use_provider, Instruction::Call { num_args: 0 });
-                Ok(())
+                let closure_to_run = if *condition { &then } else { &else_ }.address;
+                self.dup(closure_to_run);
+                DivergeControlFlow(closure_to_run)
             }
         )
     }
 
-    fn int_add(&mut self, args: Vec<Value>) -> Result<Value, String> {
-        destructure!(args, [Value::Int(summand_a), Value::Int(summand_b)], {
-            Ok((summand_a + summand_b).into())
+    fn int_add(&mut self, args: &[Pointer]) -> BuiltinResult {
+        unpack_and_later_drop!(self, args, (a: Int, b: Int), {
+            Return(self.create_int(a.value.clone() + b.value.clone()))
```

Comment:
Sadly, we do. Otherwise, the compiler complains because addition from Rust's [`Add` trait](https://doc.rust-lang.org/std/ops/trait.Add.html) has the signature `fn add(self, rhs: Rhs)`.